File: go/lib/builtins.go
package lib

import "fmt"

func builtinPrint(args []Value) Value {
	if len(args) != 1 {
		panic("print")
	}
	fmt.Println(args[0])
	return TheUndef
}

func builtinAdd(args []Value) Value {
	if len(args) == 0 {
		panic("no args")
	}
	var result int = 0
	for _, arg := range args {
		result += arg.(*Integer).RawValue
	}
	return &Integer{result}
}

func builtinSub(args []Value) Value {
	if len(args) == 0 {
		panic("no args")
	}
	var result int = args[0].(*Integer).RawValue
	for i, arg := range args {
		if i == 0 {
			continue
		}
		result -= arg.(*Integer).RawValue
	}
	return &Integer{result}
}

func builtinMul(args []Value) Value {
	if len(args) == 0 {
		panic("no args")
	}
	var result int = 1
	for _, arg := range args {
		result *= arg.(*Integer).RawValue
	}
	return &Integer{result}
}

func builtinDiv(args []Value) Value {
	if len(args) == 0 {
		panic("no args")
	}
	var result int = args[0].(*Integer).RawValue
	for i, arg := range args {
		if i == 0 {
			continue
		}
		result /= arg.(*Integer).RawValue
	}
	return &Integer{result}
}

func builtinEq(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	lhs := args[0].(*Integer).RawValue
	rhs := args[1].(*Integer).RawValue
	return MakeBoolean(lhs == rhs)
}

func builtinLt(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	lhs := args[0].(*Integer).RawValue
	rhs := args[1].(*Integer).RawValue
	return MakeBoolean(lhs < rhs)
}

func builtinLte(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	lhs := args[0].(*Integer).RawValue
	rhs := args[1].(*Integer).RawValue
	return MakeBoolean(lhs <= rhs)
}

func builtinGt(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	lhs := args[0].(*Integer).RawValue
	rhs := args[1].(*Integer).RawValue
	return MakeBoolean(lhs > rhs)
}

func builtinGte(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	lhs := args[0].(*Integer).RawValue
	rhs := args[1].(*Integer).RawValue
	return MakeBoolean(lhs >= rhs)
}

func builtinAnd(args []Value) Value {
	var result bool = true
	for _, arg := range args {
		result = result && arg.Boolean()
	}
	return MakeBoolean(result)
}

func builtinOr(args []Value) Value {
	var result bool = false
	for _, arg := range args {
		result = result || arg.Boolean()
	}
	return MakeBoolean(result)
}

func builtinNot(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	return MakeBoolean(!args[0].(*Boolean).RawValue)
}

func builtinEqCheck(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	return MakeBoolean(args[0] == args[1])
}

func builtinEqualCheck(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	return MakeBoolean(args[0].DeepEquals(args[1]))
}

func builtinNullCheck(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	_, ok := args[0].(*Null)
	return MakeBoolean(ok)
}

func builtinPairCheck(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	_, ok := args[0].(*Pair)
	return MakeBoolean(ok)
}

func builtinBooleanCheck(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	_, ok := args[0].(*Boolean)
	return MakeBoolean(ok)
}

func builtinIntegerCheck(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	_, ok := args[0].(*Integer)
	return MakeBoolean(ok)
}

func builtinSymbolCheck(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	_, ok := args[0].(*Symbol)
	return MakeBoolean(ok)
}

func builtinCons(args []Value) Value {
	if len(args) != 2 {
		panic("args")
	}
	return &Pair{args[0], args[1]}
}

func builtinCar(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	return args[0].(*Pair).Car
}

func builtinCdr(args []Value) Value {
	if len(args) != 1 {
		panic("args")
	}
	return args[0].(*Pair).Cdr
}

type nativeFunction struct {
	name string
	fun  func(args []Value) Value
}

func (fun *nativeFunction) Boolean() bool {
	panic("not boolean")
}

func (fun *nativeFunction) String() string {
	return fun.name
}

func (fun *nativeFunction) DeepEquals(other Value) bool {
	return fun == other
}

func (fun *nativeFunction) Apply(args []Value) Value {
	return fun.fun(args)
}

var builtinMap = map[string]FunctionValue{
	"print":    &nativeFunction{"print", builtinPrint},
	"+":        &nativeFunction{"+", builtinAdd},
	"-":        &nativeFunction{"-", builtinSub},
	"*":        &nativeFunction{"*", builtinMul},
	"/":        &nativeFunction{"/", builtinDiv},
	"=":        &nativeFunction{"=", builtinEq},
	"<":        &nativeFunction{"<", builtinLt},
	"<=":       &nativeFunction{"<=", builtinLte},
	">":        &nativeFunction{">", builtinGt},
	">=":       &nativeFunction{">=", builtinGte},
	"and":      &nativeFunction{"and", builtinAnd},
	"or":       &nativeFunction{"or", builtinOr},
	"not":      &nativeFunction{"not", builtinNot},
	"eq?":      &nativeFunction{"eq?", builtinEqCheck},
	"equal?":   &nativeFunction{"equal?", builtinEqualCheck},
	"null?":    &nativeFunction{"null?", builtinNullCheck},
	"pair?":    &nativeFunction{"pair?", builtinPairCheck},
	"boolean?": &nativeFunction{"boolean?", builtinBooleanCheck},
	"integer?": &nativeFunction{"integer?", builtinIntegerCheck},
	"symbol?":  &nativeFunction{"symbol?", builtinSymbolCheck},
	"cons":     &nativeFunction{"cons", builtinCons},
	"car":      &nativeFunction{"car", builtinCar},
	"cdr":      &nativeFunction{"cdr", builtinCdr},
}

File: rust/src/builtins.rs
use std::rc::Rc;

use anyhow::anyhow;
use anyhow::bail;
use anyhow::Result;

use crate::data::Value;
use crate::environment::Env;

fn builtin_print(args: &[Value]) -> Result<Value> {
    if args.len() != 1 {
        bail!("print: Invalid number of arguments");
    }
    println!("{}", args[0]);
    Ok(Value::Undef)
}

fn builtin_and(args: &[Value]) -> Result<Value> {
    let mut result = true;
    for value in args {
        result &= value.bool();
    }
    Ok(Value::Boolean(result))
}

fn builtin_or(args: &[Value]) -> Result<Value> {
    let mut result = false;
    for value in args {
        result |= value.bool();
    }
    Ok(Value::Boolean(result))
}

fn builtin_not(args: &[Value]) -> Result<Value> {
    if args.len() != 1 {
        bail!("not: Invalid number of arguments");
    }
    let value = &args[0];
    Ok(Value::Boolean(!value.bool()))
}

fn builtin_cons(args: &[Value]) -> Result<Value> {
    if args.len() != 2 {
        bail!("cons: Invalid number of arguments");
    }
    let (car, cdr) = (&args[0], &args[1]);
    Ok(Value::Pair(car.clone().into(), cdr.clone().into()))
}

fn builtin_car(args: &[Value]) -> Result<Value> {
    if args.len() != 1 {
        bail!("car: Invalid number of arguments");
    }
    let (car, _) = args[0].as_pair()?;
    let value = car.get();
    Ok(value)
}

fn builtin_cdr(args: &[Value]) -> Result<Value> {
    if args.len() != 1 {
        bail!("cdr: Invalid number of arguments");
    }
    let (_, cdr) = args[0].as_pair()?;
    let value = cdr.get();
    Ok(value)
}

fn builtin_eq(args: &[Value]) -> Result<Value> {
    if args.len() != 2 {
        bail!("=: Invalid number of arguments");
    }
    let (a, b) = (args[0].as_integer()?, args[1].as_integer()?);
    Ok(Value::Boolean(a == b))
}

fn builtin_lt(args: &[Value]) -> Result<Value> {
    if args.len() != 2 {
        bail!("<: Invalid number of arguments");
    }
    let (a, b) = (args[0].as_integer()?, args[1].as_integer()?);
    Ok(Value::Boolean(a < b))
}

fn builtin_lte(args: &[Value]) -> Result<Value> {
    if args.len() != 2 {
        bail!("<=: Invalid number of arguments");
    }
    let (a, b) = (args[0].as_integer()?, args[1].as_integer()?);
    Ok(Value::Boolean(a <= b))
}

fn builtin_gt(args: &[Value]) -> Result<Value> {
    if args.len() != 2 {
        bail!(">: Invalid number of arguments");
    }
    let (a, b) = (args[0].as_integer()?, args[1].as_integer()?);
    Ok(Value::Boolean(a > b))
}

fn builtin_gte(args: &[Value]) -> Result<Value> {
    if args.len() != 2 {
        bail!(">=: Invalid number of arguments");
    }
    let (a, b) = (args[0].as_integer()?, args[1].as_integer()?);
    Ok(Value::Boolean(a >= b))
}

fn builtin_add(args: &[Value]) -> Result<Value> {
    let mut result = 0;
    for value in args {
        result += value.as_integer()?;
    }
    Ok(Value::Integer(result))
}

fn builtin_sub(args: &[Value]) -> Result<Value> {
    let mut result = args
        .first()
        .ok_or(anyhow!("-: Invalid number of arguments"))?
        .as_integer()?;
    for value in &args[1..] {
        result -= value.as_integer()?;
    }
    Ok(Value::Integer(result))
}

fn builtin_mul(args: &[Value]) -> Result<Value> {
    let mut result = 1;
    for value in args {
        result *= value.as_integer()?;
    }
    Ok(Value::Integer(result))
}

fn builtin_div(args: &[Value]) -> Result<Value> {
    let mut result = args
        .first()
        .ok_or(anyhow!("/: Invalid number of arguments"))?
        .as_integer()?;
    for value in &args[1..] {
        result /= value.as_integer()?;
    }
    Ok(Value::Integer(result))
}

fn builtin_eq_check(args: &[Value]) -> Result<Value> {
    if args.len() != 2 {
        bail!(">: Invalid number of arguments");
    }
    let (lhs, rhs) = (&args[0], &args[1]);
    Ok(Value::Boolean(lhs == rhs))
}

pub fn install(env: &Rc<Env>) {
    for (name, func) in [
        ("print", builtin_print as fn(&[Value]) -> _),
        ("and", builtin_and),
        ("or", builtin_or),
        ("not", builtin_not),
        ("=", builtin_eq),
        ("<", builtin_lt),
        ("<=", builtin_lte),
        (">", builtin_gt),
        (">=", builtin_gte),
        ("+", builtin_add),
        ("-", builtin_sub),
        ("*", builtin_mul),
        ("/", builtin_div),
        ("eq?", builtin_eq_check),
        ("cons", builtin_cons),
        ("car", builtin_car),
        ("cdr", builtin_cdr),
    ] {
        env.ensure(name)
            .set(Value::Function(Rc::new(name.to_owned()), Rc::new(func)));
    }
}
